[dynamo] Handle general tensor attributes with a getattr proxy node

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #91840

**Background:** Before this PR, support in dynamo for tensor attributes (e.g. `x.H`, `x.T`, ...) need to be individually implemented one-by-one. This could potentially lead to errors, e.g. if the implementation in [variables/tensor.py](https://github.com/pytorch/pytorch/blob/21c7c7c72fd13f476e08b84c45cbca3ea3f41b04/torch/_dynamo/variables/tensor.py#L160) differs from the implementation from a direct call to the attribute. For attributes that were not special-cased in tensor.py, dynamo tracing would fail. This PR adds generic support for tensor attributes that return tensors without needing to specially handle them. (Notably, for x.real and x.imag, which previously weren't supported).

**In this PR:** This directly creates a proxy node for a `"call_function"` node with `target=getattr`, and feeds it into wrap_fx_proxy. This will produce a TensorVariable for the attribute returned.

This also removes the implementations for H, T, mH, mT which were broken (previously `torch.relu(x.T)` would fail). They now fall back to this default implementation (for which `torch.relu(x.T)` passes).

**Further context**:

* Ed's original suggestion in [90463](https://github.com/pytorch/pytorch/pull/90463#discussion_r1043398340) is to use `torch.Tensor.H.__get__(x)`. I wasn't able to get this to work; fx compilation fails with `getset_descriptor does not have attribute __module__`. Basically, the `__module__` attribute which is available on most python attributes, is not available on `getset_descriptor` objects. (i.e., these are implemented in C++ as attributes on torch.Tensor, so they don't obey some assumptions made by fx)
* Although both tensor attributes and methods (like `x.relu()`) both go through this, this PR should only handle attributes (e.g. see the `"getset_descriptor"` in variables/tensor.py). Methods are handled already by by GetAttrVariable.
* Prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017). So: before, for some tensor x, `x.real` would work fine; but `torch.relu(x.real)` would fail.

**Testing**: added tests in test_misc.py for x.real, x.imag, x.T, x.real.T.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

davidberard98 committed:
[dynamo] Explicitly create a GetAttrVariable for non-special-cased TensorVariable attributes

This PR provides support for tensor attributes like mT, H, mH, etc. that are valid tensor attributes, but for which there doesn't already exist a special handling for in [TensorVariable.var_getattr](https://github.com/pytorch/pytorch/blob/75c652821c3357127a4fe35ccc101077a6562015/torch/_dynamo/variables/tensor.py#L132-L159). Instead of manually adding more implementations, this PR will handle any remaining ops as GetAttrVariables. See https://github.com/pytorch/pytorch/pull/91262#discussion_r1056022981 for more discussion.

Note: prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017).

So, in this PR we explicitly create the GetAttrVariable inside TensorVariable.var_getattr, and call get_fake_value() at that point.

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Explicitly create a GetAttrVariable for non-special-cased TensorVariable attributes"


This PR provides support for tensor attributes like mT, H, mH, etc. that are valid tensor attributes, but for which there doesn't already exist a special handling for in [TensorVariable.var_getattr](https://github.com/pytorch/pytorch/blob/75c652821c3357127a4fe35ccc101077a6562015/torch/_dynamo/variables/tensor.py#L132-L159). Instead of manually adding more implementations, this PR will handle any remaining ops as GetAttrVariables. See https://github.com/pytorch/pytorch/pull/91262#discussion_r1056022981 for more discussion.

Note: prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017).

So, in this PR we explicitly create the GetAttrVariable inside TensorVariable.var_getattr, and call get_fake_value() at that point.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -556,17 +556,20 @@ def call_function(
 
 class GetAttrVariable(VariableTracker):
-    def __init__(self, obj, name, **kwargs):
+    def __init__(self, obj, name, proxy=None, **kwargs):
```
I think this might a bit confused, it seems we want to override the usual ```proxy``` from ```as_proxy```, but we didn't change the other logistics, e.g, reconstruct. It may lead inconsistent behavior when graph break happens.

davidberard98 committed:
Update on "[dynamo] Explicitly create a GetAttrVariable for non-special-cased TensorVariable attributes"


This PR provides support for tensor attributes like mT, H, mH, etc. that are valid tensor attributes, but for which there doesn't already exist a special handling for in [TensorVariable.var_getattr](https://github.com/pytorch/pytorch/blob/75c652821c3357127a4fe35ccc101077a6562015/torch/_dynamo/variables/tensor.py#L132-L159). Instead of manually adding more implementations, this PR will handle any remaining ops as GetAttrVariables. See https://github.com/pytorch/pytorch/pull/91262#discussion_r1056022981 for more discussion.

Note: prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017).

So, in this PR we explicitly create the GetAttrVariable inside TensorVariable.var_getattr, and call get_fake_value() at that point.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Create GetTensorAttrVariable to handle general tensor attributes"


**Background:** Before this PR, support in dynamo for tensor attributes (e.g. `x.H`, `x.T`, ...) need to be individually implemented one-by-one. This leaves room for error, in case the implementation in [variables/tensor.py](https://github.com/pytorch/pytorch/blob/21c7c7c72fd13f476e08b84c45cbca3ea3f41b04/torch/_dynamo/variables/tensor.py#L160) differs from the value resulting from a direct call to the attribute. For attributes that were not special-cased in tensor.py, dynamo tracing would fail. This PR adds generic support for tensor attributes that return tensors without needing to specially handle them. (Notably, for x.real and x.imag, which previously weren't supported).

**In this PR:** This introduces a subclass of GetAttrVariable, called GetTensorAttrVariable. It is specifically intended to represent attributes on tensors.

Unlike a general GetAttrVariable, a variable representing a tensor (including a GetTensorAttrVariable...

[ok I got halfway through writing this and realized this is broken, e.g. x.real.H. I'm going to go look at that...]

**Further context**:

  Note: although both tensor attributes and methods (like `x.relu()`) both go through this 

This PR provides support for tensor attributes like mT, H, mH, etc. that are valid tensor attributes, but for which there doesn't already exist a special handling for in [TensorVariable.var_getattr](https://github.com/pytorch/pytorch/blob/75c652821c3357127a4fe35ccc101077a6562015/torch/_dynamo/variables/tensor.py#L132-L159). Instead of manually adding more implementations, this PR will handle any remaining ops as GetAttrVariables. See https://github.com/pytorch/pytorch/pull/91262#discussion_r1056022981 for more discussion.

Note: prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017).

So, in this PR we explicitly create the GetAttrVariable inside TensorVariable.var_getattr, and call get_fake_value() at that point.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Create GetTensorAttrVariable to handle general tensor attributes"


**Background:** Before this PR, support in dynamo for tensor attributes (e.g. `x.H`, `x.T`, ...) need to be individually implemented one-by-one. This leaves room for error, in case the implementation in [variables/tensor.py](https://github.com/pytorch/pytorch/blob/21c7c7c72fd13f476e08b84c45cbca3ea3f41b04/torch/_dynamo/variables/tensor.py#L160) differs from the value resulting from a direct call to the attribute. For attributes that were not special-cased in tensor.py, dynamo tracing would fail. This PR adds generic support for tensor attributes that return tensors without needing to specially handle them. (Notably, for x.real and x.imag, which previously weren't supported).

**In this PR:** This directly creates a 

[ok I got halfway through writing this and realized this is broken, e.g. x.real.H. I'm going to go look at that...]

**Further context**:

  Note: although both tensor attributes and methods (like `x.relu()`) both go through this 

This PR provides support for tensor attributes like mT, H, mH, etc. that are valid tensor attributes, but for which there doesn't already exist a special handling for in [TensorVariable.var_getattr](https://github.com/pytorch/pytorch/blob/75c652821c3357127a4fe35ccc101077a6562015/torch/_dynamo/variables/tensor.py#L132-L159). Instead of manually adding more implementations, this PR will handle any remaining ops as GetAttrVariables. See https://github.com/pytorch/pytorch/pull/91262#discussion_r1056022981 for more discussion.

Note: prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017).

So, in this PR we explicitly create the GetAttrVariable inside TensorVariable.var_getattr, and call get_fake_value() at that point.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Handle general tensor attributes with a getattr proxy node"


**Background:** Before this PR, support in dynamo for tensor attributes (e.g. `x.H`, `x.T`, ...) need to be individually implemented one-by-one. This leaves room for error, in case the implementation in [variables/tensor.py](https://github.com/pytorch/pytorch/blob/21c7c7c72fd13f476e08b84c45cbca3ea3f41b04/torch/_dynamo/variables/tensor.py#L160) differs from the value resulting from a direct call to the attribute. For attributes that were not special-cased in tensor.py, dynamo tracing would fail. This PR adds generic support for tensor attributes that return tensors without needing to specially handle them. (Notably, for x.real and x.imag, which previously weren't supported).

**In this PR:** This directly creates a proxy node for a `"call_function"` node with `target=getattr`, and feeds it into wrap_fx_proxy. This will produce a TensorVariable for the attribute returned.

This also removes the implementations for H, T, mH, mT which were broken (previously `torch.relu(x.T)` would fail). They now fall back to this default implementation (for which `torch.relu(x.T)` passes).

**Further context**:

* Ed's original suggestion in [90463](https://github.com/pytorch/pytorch/pull/90463#discussion_r1043398340) is to use `torch.Tensor.H.__get__(x)`. I wasn't able to get this to work; fx compilation fails with `getset_descriptor does not have attribute __module__`. Basically, the `__module__` attribute which is available on most python attributes, is not available on `getset_descriptor` objects.
* Note: although both tensor attributes and methods (like `x.relu()`) both go through this, this PR should only handle attributes (e.g. see the `"getset_descriptor"` in variables/tensor.py).
* Note: prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017).


**Testing**: added tests in test_misc.py for x.real, x.imag, x.T, x.real.T.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Handle general tensor attributes with a getattr proxy node"


**Background:** Before this PR, support in dynamo for tensor attributes (e.g. `x.H`, `x.T`, ...) need to be individually implemented one-by-one. This could potentially lead to errors, e.g. if the implementation in [variables/tensor.py](https://github.com/pytorch/pytorch/blob/21c7c7c72fd13f476e08b84c45cbca3ea3f41b04/torch/_dynamo/variables/tensor.py#L160) differs from the implementation from a direct call to the attribute. For attributes that were not special-cased in tensor.py, dynamo tracing would fail. This PR adds generic support for tensor attributes that return tensors without needing to specially handle them. (Notably, for x.real and x.imag, which previously weren't supported).

**In this PR:** This directly creates a proxy node for a `"call_function"` node with `target=getattr`, and feeds it into wrap_fx_proxy. This will produce a TensorVariable for the attribute returned.

This also removes the implementations for H, T, mH, mT which were broken (previously `torch.relu(x.T)` would fail). They now fall back to this default implementation (for which `torch.relu(x.T)` passes).

**Further context**:

* Ed's original suggestion in [90463](https://github.com/pytorch/pytorch/pull/90463#discussion_r1043398340) is to use `torch.Tensor.H.__get__(x)`. I wasn't able to get this to work; fx compilation fails with `getset_descriptor does not have attribute __module__`. Basically, the `__module__` attribute which is available on most python attributes, is not available on `getset_descriptor` objects. (i.e., these are implemented in C++ as attributes on torch.Tensor, so they don't obey some assumptions made by fx)
* Although both tensor attributes and methods (like `x.relu()`) both go through this, this PR should only handle attributes (e.g. see the `"getset_descriptor"` in variables/tensor.py). Methods are handled already by by GetAttrVariable.
* Prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017). So: before, for some tensor x, `x.real` would work fine; but `torch.relu(x.real)` would fail.

**Testing**: added tests in test_misc.py for x.real, x.imag, x.T, x.real.T.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
getattr is a good compromise, I like it.

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -199,6 +165,40 @@ def var_getattr(self, tx, name):
+                # as `method_descriptor`s)
+                if not (
+                    str(getattr(type(static_attr), "__name__", None))
+                    == "getset_descriptor"
```
ugh, does it have to be a string comparison here?

ezyang replied:
I'd also be OK if we just have a hardcoded list of Tensor properties here too

davidberard98 replied:
found it at `types.GetSetDescriptorType`

davidberard98 committed:
Update on "[dynamo] Handle general tensor attributes with a getattr proxy node"


**Background:** Before this PR, support in dynamo for tensor attributes (e.g. `x.H`, `x.T`, ...) need to be individually implemented one-by-one. This could potentially lead to errors, e.g. if the implementation in [variables/tensor.py](https://github.com/pytorch/pytorch/blob/21c7c7c72fd13f476e08b84c45cbca3ea3f41b04/torch/_dynamo/variables/tensor.py#L160) differs from the implementation from a direct call to the attribute. For attributes that were not special-cased in tensor.py, dynamo tracing would fail. This PR adds generic support for tensor attributes that return tensors without needing to specially handle them. (Notably, for x.real and x.imag, which previously weren't supported).

**In this PR:** This directly creates a proxy node for a `"call_function"` node with `target=getattr`, and feeds it into wrap_fx_proxy. This will produce a TensorVariable for the attribute returned.

This also removes the implementations for H, T, mH, mT which were broken (previously `torch.relu(x.T)` would fail). They now fall back to this default implementation (for which `torch.relu(x.T)` passes).

**Further context**:

* Ed's original suggestion in [90463](https://github.com/pytorch/pytorch/pull/90463#discussion_r1043398340) is to use `torch.Tensor.H.__get__(x)`. I wasn't able to get this to work; fx compilation fails with `getset_descriptor does not have attribute __module__`. Basically, the `__module__` attribute which is available on most python attributes, is not available on `getset_descriptor` objects. (i.e., these are implemented in C++ as attributes on torch.Tensor, so they don't obey some assumptions made by fx)
* Although both tensor attributes and methods (like `x.relu()`) both go through this, this PR should only handle attributes (e.g. see the `"getset_descriptor"` in variables/tensor.py). Methods are handled already by by GetAttrVariable.
* Prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017). So: before, for some tensor x, `x.real` would work fine; but `torch.relu(x.real)` would fail.

**Testing**: added tests in test_misc.py for x.real, x.imag, x.T, x.real.T.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

davidberard98 committed:
Update on "[dynamo] Handle general tensor attributes with a getattr proxy node"


**Background:** Before this PR, support in dynamo for tensor attributes (e.g. `x.H`, `x.T`, ...) need to be individually implemented one-by-one. This could potentially lead to errors, e.g. if the implementation in [variables/tensor.py](https://github.com/pytorch/pytorch/blob/21c7c7c72fd13f476e08b84c45cbca3ea3f41b04/torch/_dynamo/variables/tensor.py#L160) differs from the implementation from a direct call to the attribute. For attributes that were not special-cased in tensor.py, dynamo tracing would fail. This PR adds generic support for tensor attributes that return tensors without needing to specially handle them. (Notably, for x.real and x.imag, which previously weren't supported).

**In this PR:** This directly creates a proxy node for a `"call_function"` node with `target=getattr`, and feeds it into wrap_fx_proxy. This will produce a TensorVariable for the attribute returned.

This also removes the implementations for H, T, mH, mT which were broken (previously `torch.relu(x.T)` would fail). They now fall back to this default implementation (for which `torch.relu(x.T)` passes).

**Further context**:

* Ed's original suggestion in [90463](https://github.com/pytorch/pytorch/pull/90463#discussion_r1043398340) is to use `torch.Tensor.H.__get__(x)`. I wasn't able to get this to work; fx compilation fails with `getset_descriptor does not have attribute __module__`. Basically, the `__module__` attribute which is available on most python attributes, is not available on `getset_descriptor` objects. (i.e., these are implemented in C++ as attributes on torch.Tensor, so they don't obey some assumptions made by fx)
* Although both tensor attributes and methods (like `x.relu()`) both go through this, this PR should only handle attributes (e.g. see the `"getset_descriptor"` in variables/tensor.py). Methods are handled already by by GetAttrVariable.
* Prior to this PR, we already returned GetAttrVariables for unsupported attrs: the parent caller would catch the NotImplementedError and fallback to returning a GetAttrVariable. But if this GetAttrVariable was ever passed into a torch.\* function (as it could quite possibly be, since most of these attrs are tensors), it would fail because its proxy node would be missing an [example_value](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1017). So: before, for some tensor x, `x.real` would work fine; but `torch.relu(x.real)` would fail.

**Testing**: added tests in test_misc.py for x.real, x.imag, x.T, x.real.T.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge
